src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                    // item).  However, they're typing something that doesn't seem to match at all
+                    // The completion list is just distracting at this point.
+                    return null;
+                }
+
+                MatchResult<VSCompletionItem>? bestMatchResult = null;
+                var moreThanOneMatchWithSamePriority = false;
+                foreach (var currentMatchResult in matchingItems)
+                {
+                    if (bestMatchResult == null)
+                    {
+                        // We had no best result yet, so this is now our best result.
+                        bestMatchResult = currentMatchResult;
+                    }
+                    else
+                    {
+                        var match = currentMatchResult.CompareTo(bestMatchResult.Value, _filterText);
+                        if (match > 0)
+                        {
+                            moreThanOneMatchWithSamePriority = false;
+                            bestMatchResult = currentMatchResult;
+                        }
+                        else if (match == 0)
+                        {
+                            moreThanOneMatchWithSamePriority = true;
+                        }
+                    }
+                }
+
+                int index;
+                bool hardSelect;
+
+                // If we had a matching item, then pick the best of the matching items and
+                // choose that one to be hard selected.  If we had no actual matching items
+                // (which can happen if the user deletes down to a single character and we
+                // include everything), then we just soft select the first item.
+                if (bestMatchResult != null)
+                {
+                    // Only hard select this result if it's a prefix match
+                    // We need to do this so that
+                    // * deleting and retyping a dot in a member access does not change the
+                    //   text that originally appeared before the dot
+                    // * deleting through a word from the end keeps that word selected
+                    // This also preserves the behavior the VB had through Dev12.
+                    hardSelect = !_hasSuggestedItemOptions && bestMatchResult.Value.EditorCompletionItem.FilterText.StartsWith(_filterText, StringComparison.CurrentCultureIgnoreCase);
+                    index = items.IndexOf(bestMatchResult.Value);
+                }
+                else
+                {
+                    index = 0;
+                    hardSelect = false;
+                }
+
+                return new(index,
+                    hardSelect ? UpdateSelectionHint.Selected : UpdateSelectionHint.SoftSelected,
+                    moreThanOneMatchWithSamePriority ? null : bestMatchResult.GetValueOrDefault().EditorCompletionItem);
```

Comment:
we need to doc why we pass null here if tehre are two with the same priority.

Reply:
The name seems to be a little misleading, since priority is just one of the criteria. Changed it to just moreThanOneMatch